// shoots-toolkit/lib/index.js
/*
module.exports = function(appOptions) {
    var Shoots = {};
    Shoots.user = require("./user")(appOptions);
    Shoots.flash = require("./flash")(appOptions);
    Shoots.cookies = require("./cookies");
    Shoots.util = require("./util");
    return Shoots;
}
*/
module.exports = function(appOptions, shootsConfig) {
    var collection = require('shoots-mongo')(appOptions).collection;
    var session = require('shoots-session')(appOptions);
    var cache = require('shoots-cache')(appOptions);
    var user = require("./user")(appOptions);
    var flash = require("./flash")(appOptions);
    var cookies = require("./cookies");
    var util = require("./util");

    if (!global.shoots) {
        global.shoots = {};
    }
    var shoots = {
        options: appOptions,
        config: shootsConfig,
        collection: collection,
        cache: cache,
        util: util,
        user: {
            unique: user.unique,
            add: user.add,
            update: user.update,
            remove: user.remove
        },
    };


    return function*(next) {
        var ctx = this;
        shoots.session = {
            getId: function*() {
                return yield session.getId(ctx);
            },
            get: function*(name) {
                return yield session.get(ctx, name);
            },
            set: function*(name, value, expireSeconds) {
                return yield session.set(ctx, name, value, expireSeconds);
            },
            remove: function*(name) {
                return session.remove(ctx, name);
            },
            clear: function*() {
                return yield session.clear(ctx);
            }
        };
        shoots.auth = {
            login: function*(username, password, remember, expireSeconds) {
                return yield user.login(ctx, username, password, remember, expireSeconds);
            },
            loginByUserId: function*(userId, remember, expireSeconds) {
                return yield user.loginByUserId(ctx, userId, remember, expireSeconds);
            },
            logout: function*() {
                return yield user.logout(ctx);
            },
            getUserId: function*() {
                return yield user.getUserId(ctx);
            },
            getUser: function*() {
                return yield user.getUser(ctx);
            }
        };
        shoots.flash = {
            set: function*(key, val) {
                return flash.set(ctx, key, val);
            },
            get: function*(key) {
                return flash.get(ctx, key);
            },
            has: function*(key) {
                return flash.has(ctx, key);
            }
        };
        shoots.cookies = {
            set: function*(key, val) {
                return yield cookies.set(ctx, key, val);
            },
            get: function*(key) {
                return yield cookies.get(ctx, key);
            }
        };
        this.shoots = shoots;
        yield next;
    }
}
